src/util/stream_generator.h

```diff
@@ -46,93 +33,124 @@ class suspend_sometimes {
- * suspension-related decisions.
- */
-template <size_t MIN_BUFFER_SIZE>
+// The promise type that backs the generator type and handles storage and
+// suspension-related decisions.
+template <size_t BUFFER_SIZE>
 class stream_generator_promise {
-  std::ostringstream _stream;
-  std::exception_ptr _exception;
+  std::array<char, BUFFER_SIZE> data_;
+  size_t currentIndex_ = 0;
+  static_assert(BUFFER_SIZE > 0, "Buffer size must be greater than zero");
+  // Temporarily store data that didn't fit into the buffer so far.
+  std::string_view overflow_;
+  std::exception_ptr exception_;
 
  public:
-  using value_type = std::stringbuf;
+  using value_type = std::string_view;
   using reference_type = value_type&;
   using pointer_type = value_type*;
   stream_generator_promise() = default;
 
-  basic_stream_generator<MIN_BUFFER_SIZE> get_return_object() noexcept;
+  basic_stream_generator<BUFFER_SIZE> get_return_object() noexcept;
 
   constexpr std::suspend_always initial_suspend() const noexcept { return {}; }
   constexpr std::suspend_always final_suspend() const noexcept { return {}; }
 
-  /**
-   * Handles values passed using co_yield and stores their respective
-   * string representations inside a buffer.
-   *
-   * @tparam T The Type being passed
-   * @param value The value being passed via co_yield
-   * @return Whether or not the coroutine should get suspended (currently based
-   * on isBufferLargeEnough()), wrapped inside a suspend_sometimes class.
-   */
-  CPP_template(typename S)(requires ad_utility::Streamable<S>) suspend_sometimes
-      yield_value(const S& value) noexcept {
-    // _stream appends its result to _value
-    _stream << value;
-    return suspend_sometimes{isBufferLargeEnough()};
+  // Handles strings passed using co_yield and copies them to the aggregated
+  // buffer until that buffer is full. If the buffer is too small to fit the
+  // value in its entirety, the coroutine is suspended (and thus the value kept
+  // alive) until the buffer has been consumed and has sufficient capacity. In
+  // this case the buffer will be filled to maximum capacity, so eventually
+  // every bit of `value` is written, then the coroutine will be resumed.
+  suspend_sometimes yield_value(std::string_view value) noexcept {
+    if (isBufferLargeEnough(value)) {
+      std::memcpy(data_.data() + currentIndex_, value.data(), value.size());
+      currentIndex_ += value.size();
+      overflow_ = {};
+      // Only suspend if we reached the maximum capacity exactly.
+      return suspend_sometimes{currentIndex_ == BUFFER_SIZE};
+    }
+    size_t fittingSize = BUFFER_SIZE - currentIndex_;
+    std::memcpy(data_.data() + currentIndex_, value.data(), fittingSize);
+    currentIndex_ = BUFFER_SIZE;
+    overflow_ = value.substr(fittingSize);
+    return suspend_sometimes{true};
+  }
+
+  // Overload so we can also pass char values.
+  suspend_sometimes yield_value(char value) noexcept {
```

Comment:
could use  `ad_utility::SimilarTo<char> auto` to make even less conversions work.